### PR TITLE
Trigger events when the value is prepared

### PR DESCRIPTION
**Pull request by williamsandy**
Port change from pull request for com_fields : https://github.com/joomla-projects/custom-fields/pull/90

### CODE REVIEW

**Comment by laoneo** [comment]
Please rename the events to onDPFieldsBeforePrepare and onDPFieldsAfterPrepare, otherwise we get in conflict with com_fields.

File: com_dpfields/admin/helpers/dpfields.php [hunk 1]
```diff
@@ -187,6 +197,14 @@ public static function getFields ($context, $item = null, $prepareValue = false)
 						));
 					}
 
+					/*
+					 * On after field render
+					 * Event allow plugins to modfify the output of the prepared field
+					 */
+					if ($value)
```
Comment (laoneo) [inline com_dpfields/admin/helpers/dpfields.php]: Should we not fire the event here under any circumstances, event when the $value is empty?

Comment (williamsandy) [inline com_dpfields/admin/helpers/dpfields.php]: No I think the value should not be empty. Because  I can see some plugins using this event will not even check if it is empty!
I think in general it should not be empty.

Comment (laoneo) [inline com_dpfields/admin/helpers/dpfields.php]: But IMO if you have a _before_ event also a _after_ event should be triggered. Plugin developers will expect that. This needs also to be changed in com_fields.

Comment (williamsandy) [inline com_dpfields/admin/helpers/dpfields.php]: I think it should be the caller checking to see if the value is not empty and not left to the callee. what do you think?

Comment (williamsandy) [inline com_dpfields/admin/helpers/dpfields.php]: That is true but at the moment the before event has the same check as the after event and so it makes no difference.

Comment (laoneo) [inline com_dpfields/admin/helpers/dpfields.php]: Good spot, should be sent out despite the value can be empty there too.

**Comment by williamsandy** [comment]
I have commited the changes

**Comment by laoneo** [comment]
Now you messed up the PR completely.

**Comment by williamsandy** [comment]
On a different note I have noticed that if the item parameter  is passed in as an array that the following line does not get set by the item language value
line 104: 

`self::$fieldsCache->setState('filter.language', array('*', isset($item->language) ? $item->language : JFactory::getLanguage()->getTag()));`

This is because the explicit cast of the item parameter to an object is done after this on line 109:

```
if (is_array($item))
        {
            $item = (object) $item;
        }

```

so in short filter.langue is alsways set to the languge defined by Joomla and not by the item/article if item is passed in as an array.

This is the same for com_fields aslo. Is this done by design?

**Comment by laoneo** [comment]
Please open an issue for it.

**Comment by williamsandy** [comment]
> Now you messed up the PR completely.

How I just added another commit file with the changes we discussed?

**Comment by laoneo** [comment]
https://github.com/Digital-Peak/DPFields/pull/39/commits/50a8badacd0f16ee9f06c7a3fcece475600fa3f9.

**Comment by williamsandy** [comment]
Give me one sec.

**Comment by williamsandy** [comment]
updated

**Comment by laoneo** [comment]
Sorry but something is very wrong now. Why not just reverting the last two commits' and then doing it right?
